File: Editorial/157_Read_N_Characters_Given_Read4/Java/2_Speed_Up_No_Internal_Buffer.java
public class Solution extends Reader4 {
    public int read(char[] buf, int n) {
        int copiedChars = 0;
        int readChars = 4;
        int remainingChars = n;

        // While there are at least 4 characters remaining to be read and the last
        // call to readChars returned 4 characters, write directly to buf.
        while (remainingChars >= 4 && readChars == 4) {
            char[] tempBuf = new char[4];
            readChars = read4(tempBuf);
            for (int i = 0; i < readChars; i++) {
                buf[copiedChars + i] = tempBuf[i];
            }
            copiedChars += readChars;
        }

        // If there are between 1 and 3 characters that we still want to read and
        // readChars was not 0 last time we called read4, then create a buffer
        // for just this one call so we can ensure buf does not overflow.
        if (remainingChars > 0 && readChars != 0) {
            char[] tempBuf = new char[4];
            readChars = read4(tempBuf);

            for (int i = 0; i < Math.min(remainingChars, readChars); i++) {
                buf[copiedChars++] = tempBuf[i];
            }
        }

        return Math.min(n, copiedChars);
    }
}

File: Editorial/157_Read_N_Characters_Given_Read4/Python3/2_Speed_Up_No_Internal_Buffer.py
class Solution:
    def read(self, buf, n):
        """
        :type buf: Destination buffer (List[str])
        :type n: Number of characters to read (int)
        :rtype: The number of actual characters read (int)
        """
        copiedChars = 0
        readChars = 4
        remainingChars = n

        # While there are at least 4 characters remaining to be read and the last
        # call to readChars returned 4 characters, write directly to buf.
        while remainingChars >= 4 and readChars == 4:
            buf4 = [""] * 4
            readChars = read4(buf4)
            for i in range(readChars):
                buf[copiedChars + i] = buf4[i]
            copiedChars += readChars

        # If there are between 1 and 3 characters that we still want to read and
        # readChars was not 0 last time we called read4, then create a buffer
        # for just this one call so we can ensure buf does not overflow.
        if remainingChars > 0 and readChars != 0:
            buf4 = [""] * 4
            readChars = read4(buf4)

            for i in range(min(remainingChars, readChars)):
                buf[copiedChars + i] = buf4[i]
            copiedChars += readChars

        return min(n, copiedChars)
